### ISSUE TITLE
Git Plugin

### ISSUE THREAD

**Issue by wottpal** [k-next/plugins#2]
Would be great to have only one bulletproof Git-solution for Kirby 3. I already ported the core-functionality of @pascalmh's [gcapc-plugin](https://github.com/blankogmbh/kirby-git-commit-and-push-content) to the new hooks-syntax and added a git history-field. Future ideas would be:

- @rasteiner already said he would like to PR _commaship_ compatibility 🎊
- Web Hooks
- Bundled Commits (e.g. with cronjobs and a v3 version of the [Kirby Queue](https://github.com/sebsel/queue-for-kirby))
- More Control in the Panel

Repo: http://github.com/wottpal/kirby-git (would happily add collaborators to the repo)

<img width="1139" alt="gitlog-field" src="https://user-images.githubusercontent.com/3930150/43314205-8af68950-9192-11e8-97ee-4986264b40be.png">

**Comment by bastianallgeier** [comment]
Related: https://github.com/jenstornell/kirby-plugins/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+git

**Comment by rasteiner** [comment]
What I would want from a git plugin is the ability to rollback changes, ideally in a non destructive way. 
Turns out it isn't this simple with git:

`checkout` detaches the head, so if you go back and then change something you mess up the repo. 
`revert` goes into a kind of interactive flow, which doesn't seem to be right for a web interface. 
`reset --hard` works but deletes the commits you want to rollback (so if you rollback by error you can't restore the commits). 

ideally there would be a command that sets working directory to how the repo was at a current commit, without changing head, adding or destroying commits. My git knowledge is too limited to get this working.

**Comment by wottpal** [comment]
If you would think of a "rollback of changes" not as a rollback in git-terminology but more of a part of the whole history then why not just make a new commit which reverts the content to the desired old one by hand?

**Comment by wottpal** [comment]
I just had an idea how to implement this but I have no clue on how to update content of other fields from a plugin.. 🤷‍♂️

**Comment by rasteiner** [comment]
> why not just make a new commit which reverts the content to the old one by hand?

How would that look like? Because that's kinda what I'm not getting to work. :)

it would need to discard tracked and untracked changes in the working directory and then checkout an old commit (also by deleting new files), the create a new commit that doesn't detach HEAD. 🤷‍♂️


>no clue on how to update content of other fields from a plugin..
why would you want to do this?

**Comment by wottpal** [comment]
Ok I will just describe what's in my head: A section/field plugin which can be added in each pages blueprint to enable revisions for it.* In the `index.php` I would use the [GitHelper](https://github.com/wottpal/kirby-git/blob/b80a1e3a12a5e1ba4998b1b30b131ae46eddac8e/index.php#L5) to execute some git-commands:

1. `git log --follow -- content/../current-page.txt` which shows all revisions `X` where the file was changed
2. `git show HEAD~${X}:content/../current-page.txt` for each of these revisions would give me the full yaml-content for each state

Then I would decode the YAML and pass this to my Vue-Component in `index.js` which shows a list of all revisions for the current page in a list. When the user clicks on one of the revisions its content is applied to all fields. Then the user can either *Save* or *Revert* with the core-buttons if he likes the changes or not. 


*(Maybe in the future there is a more general solution detached from the blueprints somewhere in the Panel UI)

**Comment by rasteiner** [comment]
ok, got you. 
This would work for field contents on single pages (might cause problems when the field contents are somehow referencing a file that in the meantime was deleted). But still a very nice feature. :)

As far as I know there is no "official" way to interact with other fields. 
But there are at least 2 hacky ways I can think of 😁 

1. We could overwrite the fields section and replace it with a new one that contains the original (it's very hacky but I've done it before). Once we have this "wrapper", I guess we could access [its data](https://github.com/k-next/kirby/blob/master/panel/src/components/Sections/FieldsSection.vue#L29) once it has fetched it from the api. Then we could give it new data and in theory it should update its children. 
2. we could write a mixin that is added to all components, the mixin would then check if a component is a field and, if yes, it would register it somewhere we have access to (and unregister it once it's destroyed). This way we would have a way to access them.

**Comment by bastianallgeier** [comment]
@wottpal I tried to follow the discussion and find out the scenario for the event, but maybe you could write a bit more about it. What do you want to achieve exactly?

**Comment by rasteiner** [comment]
@bastianallgeier 
the feature would be anything that allows us to control the "values" of fields sections from outside of them.

The fields sections would listen for this event and, when fired, take the event argument to patch their `values` (and trigger the save bar, I guess). 


@wottpal
you could start with this (it's the hacky method to replace the fields section to add functionality I've described above):

```js

panel.plugin('maybe/core', {
  use: [
    function (Vue) {
      const original = Vue.options.components["kirby-fields-section"]
      Vue.component('kirby-fields-section', {
        extends: original,
        created() {
          this.$events.$on('values-push', this.valuesPush)
        },
        destroyed() {
          this.$events.$off('values-push', this.valuesPush)
        },
        methods: {
          valuesPush(values) {
            let changes = false
            for (let k in values) {
              if (k in this.values) {
                this.values[k] = values[k]
                changes = true
              }
            }

            if (changes) {
              this.input(this.values)
            }
          }
        }
      })
    }
  ]
});
``` 

you can now use `this.$events.$emit('values-push', {text: 'Lorem Ipsum'})` in any component :)

**Comment by wottpal** [comment]
> @wottpal I tried to follow the discussion and find out the scenario for the event, but maybe you could write a bit more about it. What do you want to achieve exactly?

![git revisions](https://user-images.githubusercontent.com/3930150/43989999-23d4112e-9d55-11e8-9a27-81cb99c418c0.gif)

[Code](http://github.com/wottpal/kirby-git) speaks more than thousand words 😅 So I've just implemented a proof-of-concept of what I was describing in https://github.com/k-next/plugins/issues/2#issuecomment-408397408. 

To bring this from a proof-of-concept to a useful plugin with good UX I still need some help @bastianallgeier:

* Having @rasteiner's listeners in the core would be 🔥
* Fix of https://github.com/k-next/kirby/issues/574
* Possibility to listen when the Save/Revert bar pops up & hides in Vue
* When I decode a content file with `Kirby\Data\Yaml::decode($content)` the field-names are transformed like `Title-1`, is there a function which gets them back to `title_1` or should I do this by hand?

**Comment by rasteiner** [comment]
There might be a problem with the way you decode content files. Those aren't stored as yaml, but a ["Kirby txt files"](https://github.com/k-next/kirby/blob/master/src/Data/Txt.php). Notice how the single fields are separated from each other with 4 dashes(`----`) for example...

Structured fields by default save their data as yaml inside of the `Txt` format - but they don't necessarily need to: (E.g: I think `tags` fields are stored as comma separated lists, which isn't yaml). 

I think the safest way to get data out of a page from git, without checking it out, would be via Virtual Pages. Something like this (completely untested):

```php
$rawData = getRawDataFromGit();
$template = getTemplateNameFromFilename();

$content = Kirby\Data\Txt::decode($rawData);

$page = new Kirby\Cms\Page([
   'template' => $template,
   'content' => $content
]);

$dataForBrowser = Kirby\Cms\Form::for($page)->values()
```

**Comment by wottpal** [comment]
Thanks a ton for all your help!! I somehow always thought these files are yaml but of course that doesn't make much sense 🤦‍♂️🤦‍♂️

I've pushed an update which your suggestion and also added pagination ✨

Would you mind making a PR doing things in a more Vue-ish way? If I remember correctly you posted a way on Slack on how to put everything in nicely seperated `.vue` files and generate the necessary code with npm. Maybe you could set that up 😇All other improvements are also welcome, of course. But I don't mind if you don't have the time :)

**Comment by bastianallgeier** [comment]
Fantastic job!!! <3

**Comment by rasteiner** [comment]
I still have to figure out a way to compile kirby plugins that doesn't suck. The plugins where I used .vue files need to be compiled every time I make a change. Compiling takes like 3-4 seconds, not that bad if you have to do it once, but while developing (make a change, compile, refresh, test, debug, compile, refresh, test, repeat) it gets sluggish rather fast. It's not the nice "live reloading and hot swapping" we normally get with webpack dev server. 

also, kirby plugins would ideally be bundled into an "Immediately Invoked Function Expression". But that isn't supported by the vue cli and is no longer supported by `poi` (which I was using before). 
I now compile my plugins into UMD format, but that isn't ideal because it declares stuff on global space that doesn't need to be declared.

I'm afraid I have to come up with my own build logic; probably based on gulp, browserify, vueify, babel, etc... It gets complicated rather quickly. 

TLDR: I'm not ready yet :)

**Comment by wottpal** [comment]
Okay then I‘ll try to add recover files/subpages functionality first and we‘ll see if somebody found a bulletproof way of using .vue files until then :)
